### ISSUE TITLE
Security Issue: 1.0.6 introduces an additional attack surface to code injection

### ISSUE THREAD

**Issue by bluec0re** [kahanu/System.Linq.Dynamic#62]
If an application allows to insert data into a query string (e.g. to specify the property where filters should operate on) it will be vulnerable to code injection attacks.

This is possible due to a change in version 1.0.6 which removes the type whitelist for method invocations.

You'll find a more detailed description on the attack here:

https://insinuator.net/2016/10/linq-injection-from-attacking-filters-to-code-execution/

and the demo application here:

https://github.com/bluec0re/dynamic-linq-injection

**Comment by StefH** [comment]
Hello @bluec0re.

I forked you example project and replaced the library by [System.Linq.Dynamic.Core](https://github.com/stefh/System.Linq.Dynamic.Core), and it seems that this exploit is not present ?

> Unhandled Exception: System.Linq.Dynamic.Core.Exceptions.ParseException: Methods on type 'Assembly' are not accessible
>    at System.Linq.Dynamic.Core.ExpressionParser.ParseMemberAccess(Type type, Expression instance)
>    at System.Linq.Dynamic.Core.ExpressionParser.ParsePrimary()
>    at System.Linq.Dynamic.Core.ExpressionParser.ParseUnary()
>    at System.Linq.Dynamic.Core.ExpressionParser.ParseMultiplicative()
>    at System.Linq.Dynamic.Core.ExpressionParser.ParseAdditive()
>    at System.Linq.Dynamic.Core.ExpressionParser.ParseShift()
>    at System.Linq.Dynamic.Core.ExpressionParser.ParseComparison()
>    at System.Linq.Dynamic.Core.ExpressionParser.ParseLogicalAndOr()
>    at System.Linq.Dynamic.Core.ExpressionParser.ParseIn()
>    at System.Linq.Dynamic.Core.ExpressionParser.ParseConditionalAnd()
>    at System.Linq.Dynamic.Core.ExpressionParser.ParseConditionalOr()
>    at System.Linq.Dynamic.Core.ExpressionParser.ParseNullCoalescing()
>    at System.Linq.Dynamic.Core.ExpressionParser.ParseExpression()
>    at System.Linq.Dynamic.Core.ExpressionParser.Parse(Type resultType, Boolean createParameterCtor)
>    at System.Linq.Dynamic.Core.DynamicExpressionParser.ParseLambda(Boolean createParameterCtor, ParameterExpression[] parameters, Type resultType, String expression, Object[] values)
>    at System.Linq.Dynamic.Core.DynamicExpressionParser.ParseLambda(Boolean createParameterCtor, Type itType, Type resultType, String expression, Object[] values)
>    at DynamicLinq.MainClass.Main(String[] args) in C:\Projects\GitHub\forks\dynamic-linq-injection\DynamicLinq\Program.cs:line 50

Can you please double check ?

**Comment by bluec0re** [comment]
System.Linq.Dynamic.Core is not vulnerable as it does still have the type checks.

In this library, the checks were removed with the commit 37f424144c009e30a1890ff6cadd3b3424219371

**Comment by StefH** [comment]
@bluec0re Thanks for verifying.

**Comment by ryanvgates** [comment]
@bluec0re @glopesdev The checks were related to Issue #31 . I'll need to see how both the flexibility and security aspects can be achieved.

**Comment by glopesdev** [comment]
@ryanvgates how about putting the checks back in but allowing the consumer of the library to extend the type whitelist? That should be a reasonable compromise between both use cases.

@bluec0re if you can provide a breaking unit test, I can try and submit a patch/pull request for this.

**Comment by bluec0re** [comment]
Added a unit test to my project

**Comment by ryanvgates** [comment]
@bluec0re is this [the test](https://github.com/bluec0re/dynamic-linq-injection/blob/master/DynamicLinq.Test/ExploitTest.cs) that you are referring to?

**Comment by bluec0re** [comment]
Yes. Disclaimer: I've first written the test with mono and then ported to VS.

**Comment by dazinator** [comment]
Any further news on this?

**Comment by StefH** [comment]
@dazinator Just switch to [System.Linq.Dynamic.Core](https://github.com/stefh/System.Linq.Dynamic.Core)

**Comment by dazinator** [comment]
Ah lovely, cheers.
